core/src/main/java/org/infinispan/marshall/core/internal/PrimitiveExternalizer.java

```diff
@@ -0,0 +1,683 @@
+package org.infinispan.marshall.core.internal;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.jboss.marshalling.util.IdentityIntMap;
+
+final class PrimitiveExternalizer implements AdvancedExternalizer<Object> {
+
+   static final int SMALL                       = 0x100;
+   static final int MEDIUM                      = 0x10000;
+
+   // 0x03-0x0F reserved
+   static final int NULL                        = 0x00;
+   static final int BYTE_ARRAY                  = 0x01; // byte[].class
+   static final int STRING                      = 0x02; // String.class
+
+   static final int BOOLEAN_OBJ                 = 0x10; // Boolean.class
+   static final int BYTE_OBJ                    = 0x11; // ..etc..
+   static final int CHAR_OBJ                    = 0x12;
+   static final int DOUBLE_OBJ                  = 0x13;
+   static final int FLOAT_OBJ                   = 0x14;
+   static final int INT_OBJ                     = 0x15;
+   static final int LONG_OBJ                    = 0x16;
+   static final int SHORT_OBJ                   = 0x17;
+
+   static final int BOOLEAN_ARRAY               = 0x18; // boolean[].class
+   static final int CHAR_ARRAY                  = 0x19; // ..etc..
+   static final int DOUBLE_ARRAY                = 0x1A;
+   static final int FLOAT_ARRAY                 = 0x1B;
+   static final int INT_ARRAY                   = 0x1C;
+   static final int LONG_ARRAY                  = 0x1D;
+   static final int SHORT_ARRAY                 = 0x1E;
+
+   static final int OBJECT_ARRAY                 = 0x1F; // Object[]
```

Comment:
Btw, we could potentially have primitive section deal with generic arrays in case they're found by chance. Currently the primitive section deals with basic Object[] and primitive types, but JBoss Marshaller code goes beyond that to deal with any type of arrays. By doing this, we would effectively get rid of ArrayExternalizer. Nice to have potentially and I'll make sure I track it down.

To reiterate, when we know we have an array, we should call to MarshallUtil, but for those cases where the type is suddenly an array and we can't guess, we need to handle it.